pkg/template/controller/templateinstance_controller.go

```diff
@@ -132,38 +140,130 @@ func (c *TemplateInstanceController) sync(key string) error {
 		return err
 	}
 
-	instantiateErr := c.instantiate(templateInstance)
-	if instantiateErr != nil {
-		glog.V(4).Infof("TemplateInstance controller: instantiate %s returned %v", key, instantiateErr)
+	if !templateInstance.HasCondition(templateapi.TemplateInstanceWaiting, kapi.ConditionTrue) {
+		err = c.instantiate(templateInstance)
+		if err != nil {
+			glog.V(4).Infof("TemplateInstance controller: instantiate %s returned %v", key, err)
+
+			templateInstance.SetCondition(templateapi.TemplateInstanceCondition{
+				Type:    templateapi.TemplateInstanceInstantiateFailure,
+				Status:  kapi.ConditionTrue,
+				Reason:  "Failed",
+				Message: err.Error(),
+			})
+		}
 	}
 
-	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceReady)
-	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceInstantiateFailure)
+	if err == nil {
+		err = c.checkReadiness(templateInstance)
+		if err != nil {
+			glog.V(4).Infof("TemplateInstance controller: checkReadiness %s returned %v", key, err)
+
+			templateInstance.DeleteCondition(templateapi.TemplateInstanceWaiting)
```

Comment:
I thought the philosophy behind Conditions was that you just add/update them?  Should we not move the Waiting condition back to false instead of deleting it?  (obviously this requires other logic in the controller for deciding whether it needs to process a template or not..presumably that could be checking for the Sucess || Failed conditions).